core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -0,0 +1,220 @@
+    private static final int FORMAT_VERSION = 1;
+
+    /** The keystore type for a newly created keystore. */
+    private static final String NEW_KEYSTORE_TYPE = "PKCS12";
+
+    /** The algorithm used to store password for a newly created keystore. */
+    private static final String NEW_KEYSTORE_SECRET_KEY_ALGO = "PBEWithHmacSHA256AndAES_128";
+
+    /** True iff the keystore has a password needed to read. */
+    private final boolean hasPassword;
+
+    /** The type of the keystore, as passed to {@link java.security.KeyStore#getInstance(String)} */
+    private final String type;
+
+    /** A factory necessary for constructing instances of secrets in a {@link KeyStore}. */
+    private final SecretKeyFactory secretFactory;
+
+    /** A stream of the actual keystore data. */
+    private final InputStream input;
+
+    /** The loaded keystore. See {@link #loadKeystore(char[])}. */
+    private final SetOnce<KeyStore> keystore = new SetOnce<>();
+
+    /** The password for the keystore. See {@link #loadKeystore(char[])}. */
+    private final SetOnce<KeyStore.PasswordProtection> keystorePassword = new SetOnce<>();
+
+    /** The setting names contained in the loaded keystore. */
+    private final Set<String> settingNames = new HashSet<>();
+
+    private KeyStoreWrapper(boolean hasPassword, String type, String secretKeyAlgo, InputStream input) {
+        this.hasPassword = hasPassword;
+        this.type = type;
+        try {
+            secretFactory = SecretKeyFactory.getInstance(secretKeyAlgo);
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+        this.input = input;
+    }
+
+    /** Returns a path representing the ES keystore in the given config dir. */
+    static Path keystorePath(Path configDir) {
+        return configDir.resolve("elasticsearch.keystore");
+    }
+
+    /** Constructs a new keystore with the given password. */
+    static KeyStoreWrapper create(char[] password) throws Exception {
+        KeyStoreWrapper wrapper = new KeyStoreWrapper(password.length != 0, NEW_KEYSTORE_TYPE, NEW_KEYSTORE_SECRET_KEY_ALGO, null);
+        KeyStore keyStore = KeyStore.getInstance(NEW_KEYSTORE_TYPE);
+        keyStore.load(null, null);
+        wrapper.keystore.set(keyStore);
+        wrapper.keystorePassword.set(new KeyStore.PasswordProtection(password));
+        return wrapper;
+    }
+
+    /**
+     * Loads information about the Elasticsearch keystore from the provided config directory.
+     *
+     * {@link #loadKeystore(char[])} must be called before reading or writing any entries.
+     * Returns {@code null} if no keystore exists.
+     */
+    public static KeyStoreWrapper loadMetadata(Path configDir) throws IOException {
+        Path keystoreFile = keystorePath(configDir);
+        if (Files.exists(keystoreFile) == false) {
+            return null;
+        }
+        DataInputStream inputStream = new DataInputStream(Files.newInputStream(keystoreFile));
```

Comment:
can we make use of CodecUtil and write a head as well as a footer. I really want this to be checksummed given that we embed this keystore in our own file. We already have `InputStreamIndexInput` and `OutputStreamIndexOutput` so I think it should be doable?